napari/layers/utils/color_manager.py

```diff
@@ -0,0 +1,521 @@
+    ):
+        """Set a color property. This is convenience function
+
+        Parameters
+        ----------
+        color : (N, 4) array or str
+            The value for setting edge or face_color
+        n_colors : int
+            The number of colors that needs to be set. Typically len(data).
+        properties : Dict[str, np.ndarray]
+            The layer property values
+        current_properties : Dict[str, np.ndarray]
+            The layer current property values
+        """
+        # if the provided color is a string, first check if it is a key in the properties.
+        # otherwise, assume it is the name of a color
+        if is_color_mapped(color, properties):
+            self.color_properties = ColorProperties(
+                name=color,
+                values=properties[color],
+                current_value=np.squeeze(current_properties[color]),
+            )
+            if guess_continuous(properties[color]):
+                self.mode = ColorMode.COLORMAP
+            else:
+                self.mode = ColorMode.CYCLE
+        else:
+            transformed_color = transform_color_with_defaults(
+                num_entries=n_colors,
+                colors=color,
+                elem_name="color",
+                default="white",
+            )
+            colors = normalize_and_broadcast_colors(
+                n_colors, transformed_color
+            )
+            self.mode = ColorMode.DIRECT
+            self.colors = colors
+
+    def refresh_colors(
+        self,
+        properties: Dict[str, np.ndarray],
+        update_color_mapping: bool = False,
+    ):
+        """Calculate and update colors if using a cycle or color map
+        Parameters
+        ----------
+        properties : Dict[str, np.ndarray]
+           The layer properties to use to update the colors.
+        update_color_mapping : bool
+           If set to True, the function will recalculate the color cycle map
+           or colormap (whichever is being used). If set to False, the function
+           will use the current color cycle map or color map. For example, if you
+           are adding/modifying points and want them to be colored with the same
+           mapping as the other points (i.e., the new points shouldn't affect
+           the color cycle map or colormap), set update_color_mapping=False.
+           Default value is False.
+        """
+        if self.mode in [ColorMode.CYCLE, ColorMode.COLORMAP]:
+            property_name = self.color_properties.name
+            current_value = self.color_properties.current_value
+            property_values = properties[property_name]
+            self.color_properties = ColorProperties(
+                name=property_name,
+                values=property_values,
+                current_value=current_value,
+            )
+
+            if update_color_mapping is True:
+                self.contrast_limits = None
+            self.events.color_properties()
```

Comment:
Is there  a way to force event emission when setting a field? This method is here for backwards compatibility. We have previously had a `refresh_colors()` method to support the user changing properties in place and then forcing the colors to be updated. Setting the `color_properties` won't emit an event automatically because the "new" value is not different ([see the keybinding in the add_points_with_properties.py example](https://github.com/napari/napari/blob/92c104dceed51727b5017f8d8cde5d1f91e0493d/examples/add_points_with_properties.py#L45-L55)). Thoughts?

Reply:
hmm, if we need to leave this we should do `self.events.color_properties(value=self.color_properties)` for consistency with the normal event emission, but i'm not sure if we shouldn't drop this method once all the refactoring is done. Maybe make the method private for this PR and then we can try and drop in a follow-up?

Reply:
@sofroniewn , I totally agree that we should drop this method. The only time this is useful is when the user modifies a field/property in place. I think we should discourage users from modifying fields/properties in place since it bypasses the validators.